### PR TITLE
Use phase barriers to work around the CUDA driver bug

### PR DESCRIPTION
**Pull request by magnatelee**
[nv-legate/cunumeric/#374](https://github.com/nv-legate/cunumeric/issues/374) turned out to be caused by a CUDA driver bug that has been fixed in CUDA driver >= 520. This PR is a workaround that introduces a barrier at the beginning of every NCCL task whenever the system is using a buggy driver.

### CODE REVIEW

File: install.py [hunk 1]
```diff
@@ -421,6 +421,14 @@ def build_legion(
             + setup_py_flags,
             cwd=legion_python_dir,
         )
+    verbose_check_call(
```
Comment (manopapad) [inline install.py:424]: Should this be added to the `INSTALL_HEADERS` in https://github.com/StanfordLegion/legion/blob/master/runtime/runtime.mk#L1030?

Comment (magnatelee) [reply install.py:424]: I guess we should check with Mike but I don't think we actually want to expose that header, though it's necessary for the core @lightsighter any thoughts?

File: src/core/utilities/deserializer.h [hunk 2]
```diff
@@ -89,6 +89,7 @@ class TaskDeserializer : public BaseDeserializer<TaskDeserializer> {
   void _unpack(RegionField& value);
   void _unpack(OutputRegionField& value);
   void _unpack(comm::Communicator& value);
+  void _unpack(Legion::PhaseBarrier& future);
```
Comment (manopapad) [inline src/core/utilities/deserializer.h:92]: rename arg to `barrier` to match definition

File: legate/core/launcher.py [hunk 3]
```diff
@@ -901,6 +906,11 @@ def build_task(
             req.proj.add(task, req, fields, _index_task_calls)
         for future in self._future_args:
             task.add_future(future)
+        if self._insert_barrier:
+            volume = launch_domain.get_volume()
+            arrival, wait = self._runtime.get_barriers(volume)
+            task.add_future(arrival)
+            task.add_future(wait)
```
Comment (manopapad) [inline legate/core/launcher.py:913]: How come these are passed as futures and not scalars?

File: src/core/comm/comm_nccl.cu [hunk 4]
```diff
@@ -170,6 +189,20 @@ void register_tasks(Legion::Machine machine,
   }
 }
 
+static const char* MIN_VERSION_WITH_HANG_FIX = "520";
+
+bool needs_barrier()
+{
+  CHECK_NVML(nvmlInit_v2());
```
Comment (manopapad) [inline src/core/comm/comm_nccl.cu:196]: Could you add a couple of sentences here for future readers, describing the issue and workaround? It's clear by reading the code how the workaround functions, but it's not clear why it's needed.

**Review by manopapad — APPROVED** [review]
Some minor questions, otherwise looks good.

File: src/core/utilities/deserializer.h (hunk 2 above)
Comment (magnatelee) [reply src/core/utilities/deserializer.h]: 1cb2ff61bdc52cacad92ff1f2b227dc2daf85285

File: src/core/comm/comm_nccl.cu (hunk 4 above)
Comment (magnatelee) [reply src/core/comm/comm_nccl.cu]: 72cd10eb3abbfcdde9d6d0d8faab5289592e4764

File: legate/core/launcher.py (hunk 3 above)
Comment (magnatelee) [reply legate/core/launcher.py:913]: I haven't found a sane way to pack cdata objects in the argument buffer.

File: src/core/comm/comm_nccl.cu [hunk 5]
```diff
@@ -170,6 +171,27 @@ void register_tasks(Legion::Machine machine,
   }
 }
 
+bool needs_barrier()
+{
+  int32_t ver;
+  auto status = cuDriverGetVersion(&ver);
+  if (status != CUDA_SUCCESS) {
+    const char* error_string;
+    cuGetErrorString(status, &error_string),
```
Comment (manopapad) [inline src/core/comm/comm_nccl.cu:180]: Should be semicolon here instead of comma?

Comment (magnatelee) [reply src/core/comm/comm_nccl.cu]: 12aef5f92ecfcf6be0eefd0191b2e0a9e80bafc0

**Pull request merged**